### DISCUSSION TITLE
Why does user config have the highest precedence instead of the lowest?

### DISCUSSION THREAD

**Discussion by khuldraeseth** [clangd/clangd#1193] [Q&A]
From the [docs](https://clangd.llvm.org/config), I see that the order of precedence is user config > inner project > outer project. The order fragments are applied in practice matches, with user config coming last. I would expect inner > outer > user instead, to allow each project to provide its own `.clangd` that can override whatever opinionated configuration I have in my user config.

A concrete example is this very project. I have the clang-tidy check `modernize-use-trailing-return-type` enabled in my user config, but the clangd source doesn't use trailing return types. A project configuration `Diagnostics.ClangTidy.Remove: modernize-use-trailing-return-type` does not remove the check because it's applied before the check is added.

**Comment by HighCommander4** [comment]
I suspect the thinking is that project configuration is likely to be checked into the project's repository and thus shared by all developers, and individual developers may want to be able to override aspects of the shared configuration.

You can add things to user config that only apply to one project, as follows:

```
// previous fragments of config file

---

If:
  PathMatch: /path/to/project

Key: Value
```

**Reply by khuldraeseth** [reply]
It still feels backwards to me. The project configuration would be checked in because it's the proper configuration to use for the project, and individual developers overriding anything would only cause headache. Why have a user config at all, if it'll have to be edited anyway for each project?

**Reply by Katzeee** [reply]
I'm confused too. That means my global remove clang-tidy flags may remove the flags set for the project.

**Reply by emilyyyylime** [reply]
+1. This is totally backwards. I want to have access to C++20 so I have `-std=c++20` in my user flags but now I want to work on a C project and just. Can't? Because there is no way for me to remove the flags in the project config

**Reply by Freed-Wu** [reply]
Any progress about it?

**Reply by khuldraeseth** [reply]
I've been getting by with a tiny patch that [pushes project config](https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/tool/ClangdMain.cpp#L944-L945) above user config in the provider stack. No problems so far.

**Reply by Freed-Wu** [reply]
Can you provide a tiny PR for your tiny patch?

**Comment by MythreyaK** [comment]
I was wondering if a new CLI arg, say, `--new-config-hierarchy` or an environment variable could be added so that the user can opt into the new behavior which applies the configs in this order  

```
CLI settings > project-level (.clangd) > parent dirs (.clangd) > user-config (.clangd)
```

This would allow for setups where
* A "global" config can be set by users in their `~/.config/clangd/clangd.yaml` 
* Any conflicting keys in projects can be easily overridden, allowing it to be committed to version control
* Prevent unnecessary edits to the user-level config when users work on new projects 
* No impact to existing configs because this is an opt-in feature

**Reply by recursivenomad** [reply]
I like this approach too.  @HighCommander4, would you be open to this becoming a feature a request?

If backwards-compatibility is a priority and we want to avoid a new command switch, could alternatively admit a second global `passiveconfig.yaml` / `defaultconfig.yaml` or similar, which receives lowest priority in the chain.

**Reply by HighCommander4** [reply]
> @HighCommander4, would you be open to this becoming a feature a request?

As this is a design decision made by the original authors (@ilya-biryukov, @sam-mccall, @kadircet, @hokein), I would like to hear from them before considering a change.

**Reply by kadircet** [reply]
the rationale behind current precedence is; Clangd is meant to provide "user" with a personal editing experience. Hence clangd's config mechanism is meant to express that stylistic preferences, it wasn't meant as a way to "enforce" clang-tidy checks and whatnot.

Maybe we lost sight of this rationale when making new features available to configuration mechanism at times. So if you people can provide more information about what particular knobs are not working as intended with this approach, we can assess the situation better.

Regarding clang-tidy check configuration specifically, we introduced `Remove` knob because certain clang-tidy checks would trigger crashes on incomplete code (which clangd needs to deal with). I guess we didn't think thoroughly about implications on having `Add`. But in theory `.clang-tidy` files should still be preferred to configure checks. If a user chooses to enable a tidy check in clangd through user config, that sounds like they want that check to be enabled everywhere (even if a project doesn't enable it by default). So I feel like the special case mentioned in the original discussion is very specific/rare and should be addressed by enabling such tidy checks only for particular projects (preferably via `.clang-tidy` config) rather than through default clangd user config.

**Reply by recursivenomad** [reply]
Thanks for the insight, @kadircet.

I definitely see the value of `config.yaml` asserting the highest priority so clangd has system-specific configuration settings that cannot be overridden (for example, I define my compiler's `-isystem` flags, which I obviously would not want ship with my code in case others wish to user alternative compilers).

However, the use case for having an additional `defaultconfig.yaml`, or a third knob for `Add`, `Remove`, and `Default`, would be that I have stylistic preferences that I like to view projects with by default, and if the project I am inspecting does not define any stylistic preferences, then (and only then) I would like my default preferences to apply as a fall-back.  Otherwise, I believe my only option as you say above is to manually add my own `.clangd` or `.clang-tidy` file to the root of the project I am inspecting, and it would be nice if instead this was automatic.  Essentially a way to have a preference without committing that preference to a repo, should that repo not care about style (and without impacting repos that do assert a style).

Let me know if this is a helpful example, or if more specific details would be required.

**Reply by kadircet** [reply]
Thanks, I think that's helpful, but I feel it's still too focused on a concrete use case that isn't obvious.

For example, the approach you mentioned won't really resolve https://github.com/clangd/clangd/discussions/1193#discussion-4155958, sometimes not having any customizations (as in the case of llvm project) is enough to diverge from user configuration. So the user will still need to take extra action.

**Reply by recursivenomad** [reply]
Perhaps I'm not understanding the purpose of clangd then.  For context, the only llvm-project software I have installed on my computer is clangd for code completion/linting; everything else is GCC-based.  The purpose of a `.clangd` file in my repository root is to assert style for the project via the `ClangTidy:` (and other) categories.

Another example I can provide is that without a `.clangd` file and an empty `config.yaml`, most projects generate warning noise due to `UnusedIncludes: Strict` being a default setting.  I would like to have a project without a `.clangd` file to default to not generating this noise, and so it would be nice to have an overwritable default configuration which sets `UnusedIncludes: None`, but allows for projects which explicitly set `UnusedIncludes: Strict` in their `.clangd` file to still enforce that style.

But I see that [llvm/llvm-project](https://github.com/llvm/llvm-project) does not contain a `.clangd` file; however it does contain `.clang-tidy`.  What I would expect would happen in this case with a `defaultconfig.yaml` is for non-clang-tidy settings, like `UnusedIncludes`, to defer to their defaults, whereas clang-tidy defaults would be wholly ignored due to the presence of a `.clang-tidy` file.

So to me, such a "default" approach would resolve the description you linked, as the `.clang-tidy` file would wholly enforce its style in place of any `defaultconfig.yaml` clang-tidy settings, and the remaining (non-clang-tidy) `defaultconfig.yaml` settings would persist due to the llvm-project repo not asserting those other options one way or the other.

But as I said, perhaps this is not the intended way to think about clangd.  If anything feels off about my usage that I can learn from, please let me know.
